### DISCUSSION TITLE
Nullable annotation at the file level/deliberate mis-annotating functions

### DISCUSSION THREAD

**Discussion by jhudsoncedaron** [dotnet/csharplang#8170] [General] [answered]
I had previously wondered at why there's no `#nullable annotations` to go with `<Nullable>annotations</Nullable>`. We have had `#nullable enable` slowly spreading through the codebase but the exercise of setting `<Nullable>annotations</Nullable>` is still far beyond reach. I had this one finally appear in the codebase:

```c#
   string ValueOrDefault(string key, Func<string> default);
```

It really wants to be

```c#
   string? ValueOrDefault(string key, Func<string?> default);
   string ValueOrDefault(string key, Func<string?> default);
```

but that's impossible. The next best idea I had was:

```
   #nullable annotate
   string ValueOrDefault(string key, Func<string?> default);
```

That is, deliberately mis-annotating it so we don't have to deal with the fact it can't be declared; but then its own implementation can't be `#nullable enable`.

I've made this as a discussion item because these are rather brute force ideas and there's probably something better out there.

**Comment by 333fred** [comment, marked as answer]
`#nullable enable annotations` - https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#nullable-context
